support-frontend/assets/pages/[countryGroupId]/components/retryPaymentStatus.ts

```diff
@@ -0,0 +1,51 @@
+import type { StatusResponse } from 'helpers/forms/paymentIntegrations/readerRevenueApis';
+
+// ----- Setup ----- //
+const POLLING_INTERVAL_MILLIS = 3000;
+const MAX_POLLS = 10;
+
+// corrects ES5 instanceOf error for retry mechanism
+// https://dev.to/dguo/how-to-fix-instanceof-not-working-for-custom-errors-in-typescript-4amp
+class StillPendingError extends Error {
+	constructor(message: string) {
+		super(message);
+		Object.setPrototypeOf(this, StillPendingError.prototype);
+	}
+}
+
+function timeOut(milliseconds: number | undefined): Promise<void> {
+	return new Promise((resolve) => setTimeout(resolve, milliseconds));
+}
+
+export function retryPaymentStatus(
+	promiseFunction: () => Promise<StatusResponse>,
+	pollMax: number = MAX_POLLS,
+	pollInterval: number = POLLING_INTERVAL_MILLIS,
+	onRetry?: (pollCount: number, pollTimeTotal: number) => void,
+): Promise<StatusResponse> {
+	async function retryPollAndPromise(polls: number): Promise<StatusResponse> {
+		try {
+			if (polls > 0) {
+				await timeOut(pollInterval); // retry
+			}
+			const result = await promiseFunction();
+			if (result.status === 'pending') {
+				throw new StillPendingError('status is pending'); // retry on pending
+			}
+			return result; // success or failure, exit
+		} catch (error) {
+			if (polls < pollMax) {
+				if (onRetry) {
+					onRetry(polls + 1, (polls + 1) * pollInterval); // optional retry notification
+				}
+				return retryPollAndPromise(polls + 1);
+			} else {
+				if (error instanceof StillPendingError) {
+					return { status: 'pending', trackingUri: '' };
```

Comment:
Yes that is a good idea rather than leave blank. Coul donly seem to pass Uri through as the message. Is tha acceptable (with only only string needed) or there is a better way?

Reply:
Heya, yeah I think that works, since we're not using the message for anything else. I think an alternative could be to augment `StillPendingError` with a custom property [like this](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDKAuBLEIAKBTAdgE0Zg5gKIBOxA9sdOgB7xbYwnmUDeAUNNMevMQJ4BVYogBc0CLzz4A3GzYBIYGUwTiAV2DwKACgC26KGHzoxqqQBouPfkNHjJBAJTR28+RDUAHdMT0GIRuiOsm4A8gBGAFbomgB0EDyo5FrwfN6hAGba8AAWiBCWCMho9FJMFLGeyWSp3sEcnNC5+bHcvILC0AC8Vu22IQC+bENs7S4NueQA7tCY6DNFKBg4ZaQ6AERkZJ4QAITrlus58PA7IgD05zRgup4g6LFKuudt-Ov1A9DAYPDAOdDaILjTiIDIA9DQPASMCYYDoMhgxYlFYEcrEZzsRpfZQQMj3WIgMj4QGtawdRD1ThDIZAA).

Reply:
Thats perfect, just updated with ath approach, thanks.